clients/da-vinci-client/src/test/java/com/linkedin/davinci/store/rocksdb/RocksDBThrottlerTest.java

```diff
@@ -0,0 +1,134 @@
+          try {
+            throttler.throttledOperation("/test_" + threadId, new TestOperationRunnable(threadId, latch, waitTimeMS));
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        });
+      }
+    } finally {
+      TestUtils.shutdownExecutor(executorService, (numTasks * waitTimeMS) + bufferWaitMS, TimeUnit.MILLISECONDS);
+      latch.await();
+    }
+  }
+
+  /**
+   * Correctness: Testing whether only the allowedMaxOperationsInParallel are executed at any time.
+   * - allowedMaxOperationsInParallel is set to 5, number of tasks is set to 10 and each tasks will
+   *   sleep for 10 sec, so all the tasks should have finished executing in 20 seconds, but that is not
+   *   being checked here.
+   * - Once the first 5 tasks starts getting executed (ie starts sleeping), we wait for 1 second and
+   *   verify whether CountDownLatch is equal to 5 which is the remaining tasks after the first batch
+   *   of 5 tasks getting executed, to verify whether the throttler only executes a maximum of this many
+   *   number of tasks in parallel.
+   * - again check after 7 seconds (total 8 seconds) and CountDownLatch should be 5 even now as the
+   *   first set of tasks are still sleeping
+   * - again check after 3 seconds (total 11 seconds) and CountDownLatch should be 0 as the second set
+   *   of tasks should have started executing as the sleep time inside task is only 10 seconds.
+   */
+  @Test(timeOut = 25 * Time.MS_PER_SECOND)
+  public void testThrottleCorrectness() throws InterruptedException {
+    int numTasks = 10;
+    int waitTimeMS = 10 * Time.MS_PER_SECOND; // 10 secs
+    int allowedMaxOperationsInParallel = 5;
+    ExecutorService executorService = Executors.newFixedThreadPool(numTasks);
+    // Create a CountDownLatch with the number of tasks
+    CountDownLatch latch = new CountDownLatch(numTasks);
+
+    try {
+      RocksDBIngestThrottler throttler = new RocksDBIngestThrottler(allowedMaxOperationsInParallel);
+      for (int i = 0; i < numTasks; ++i) {
+        final int threadId = i;
+        executorService.submit(() -> {
+          try {
+            throttler.throttledOperation("/test_" + threadId, new TestOperationRunnable(threadId, latch, waitTimeMS));
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        });
+      }
+    } finally {
+      Thread.sleep(1 * Time.MS_PER_SECOND);
+      // First 5 tasks should be currently executing, so 5 remaining tasks
+      Assert.assertEquals(latch.getCount(), allowedMaxOperationsInParallel);
+      Thread.sleep(7 * Time.MS_PER_SECOND);
+      // First 5 tasks should still be currently executing, so 5 remaining tasks
+      Assert.assertEquals(latch.getCount(), allowedMaxOperationsInParallel);
+      Thread.sleep(3 * Time.MS_PER_SECOND);
+      // Next 5 tasks should have started executing as its more than 10 seconds, so 0 remaining tasks
+      Assert.assertEquals(latch.getCount(), 0);
```

Comment:
Relying on unconditional sleeps in multithreaded environment is a recipe for a flaky test! Please don't user sleep in tests.

I suggest to coordinate test execution directly from the test thread to make it deterministic. For example:

Create a throttler with limit 1
Submit a task to thread pool that will call throttledOperation with latch.await() as the payload.
Submit another task to thread pool that will call throttledOperation with empty lambda as the payload.
Assert that the second task is not complete.
Release the latch.
Assert that the first task is complete.
Assert that the second task is complete.
Such test verifies both correctness and liveness. You can do it with a larger number of tasks as well.